src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs

```diff
@@ -11,108 +11,178 @@
+                return _packages[requestedPackageType];
             }
             finally
             {
                 if (acquired)
                 {
-                    pushedPackage.PushLock.Release();
+                    _pushLock.Release();
                 }
             }
         }
 
-        public void Dispose()
+        private Package GetPushedPackage(PackageType requestedPackageType, ITestOutputHelper logger)
         {
+            lock (_packagesLock)
+            {
+                if (_packages.TryGetValue(requestedPackageType, out Package package))
+                {
+                    logger.WriteLine($"Package of type {requestedPackageType} has already been pushed. Using {package}.");
+                    return package;
+                }
+            }
+
+            return null;
         }
 
-        private Package CreatePackage(PackageType packageType)
+        private async Task<Package> UncachedPushAsync(PackageType requestedPackageType, PackageType packageType, ITestOutputHelper logger)
         {
-            switch (packageType)
+            await Task.Yield();
+
+            Package package = CreatePackage(packageType);
+            logger.WriteLine($"Package of type {packageType} has not been pushed yet. Pushing {package}.");
+            try
             {
-                case PackageType.SemVer2Prerelease:
-                    return Package.Create(packageType.ToString(), "1.0.0-alpha.1");
-                default:
-                    return Package.Create(packageType.ToString(), "1.0.0");
+                using (var nupkgStream = new MemoryStream(package.NupkgBytes.ToArray()))
+                {
+                    await _clients.Gallery.PushAsync(nupkgStream);
+                }
+                logger.WriteLine($"Package {package} has been pushed.");
             }
+            catch (Exception ex)
+            {
+                if (packageType == requestedPackageType)
+                {
+                    throw;
+                }
+                else
+                {
+                    logger.WriteLine(
+                        $"The package {package} failed to be pushed. Since this package was not explicitly requested " +
+                        $"by the running test, this failure will be ignored. Exception:{Environment.NewLine}{ex}");
+                    return null;
+                }
+            }
+            
+            return package;
         }
 
-        private class PushedPackage
+        public void Dispose()
         {
-            public PushedPackage(Package package)
-            {
-                PushLock = new SemaphoreSlim(1);
-                Package = package;
-                Pushed = false;
-            }
+        }
 
-            public SemaphoreSlim PushLock { get; }
-            public Package Package { get; }
-            public bool Pushed { get; private set; }
+        private IEnumerable<PackageType> GetPackageTypes(PackageType requestedPackageType)
```

Comment:
I wanted to be "correct". Enums can have integer values that do not correspond to a name.